File: services/meeting-api/meeting_api/runtime_backend.py
"""Local-vs-RunPod runtime-api backend selection.

Replaces the standalone `router` service. Router special-cased `POST /bots` /
`DELETE /bots/{platform}/{id}` with a sticky, in-memory local/RunPod split — but nothing ever
called those routes (runtime-api only exposes /containers, not /bots), so real traffic went
through router's catch-all proxy instead, which routed on the single static USE_LOCAL_RESOURCE
env var only. There was no actual per-bot load balancing happening.

This picks a backend once per meeting at spawn time based on current local-bot occupancy vs
LOCAL_BOT_THRESHOLD, persists the choice in meeting.data (survives restarts — the in-memory map
didn't), and resolves it back out for every later container operation on that meeting.
"""

from sqlalchemy import and_, func, select
from sqlalchemy.ext.asyncio import AsyncSession

from .config import LOCAL_BACKEND_URL, LOCAL_BOT_THRESHOLD, RUNPOD_BACKEND_URL, USE_LOCAL_RESOURCE
from .models import Meeting

ACTIVE_STATUSES = ["requested", "joining", "awaiting_admission", "active"]


async def choose_backend_for_spawn(db: AsyncSession) -> str:
    """Pick "local" or "runpod" for a *new* bot about to be spawned."""
    if not USE_LOCAL_RESOURCE:
        return "runpod"

    count_stmt = select(func.count()).select_from(Meeting).where(
        and_(
            Meeting.platform != "browser_session",
            Meeting.status.in_(ACTIVE_STATUSES),
            Meeting.data["runtime_backend"].astext == "local",
        )
    )
    local_count = int((await db.execute(count_stmt)).scalar() or 0)
    return "local" if local_count < LOCAL_BOT_THRESHOLD else "runpod"


def backend_url(backend: str) -> str:
    return LOCAL_BACKEND_URL if backend == "local" else RUNPOD_BACKEND_URL


def meeting_backend_url(meeting: Meeting) -> str:
    """Resolve the backend URL already decided for an existing meeting. Falls back to the
    static USE_LOCAL_RESOURCE flag only for meetings created before this field existed."""
    backend = (meeting.data or {}).get("runtime_backend")
    if backend is None:
        backend = "local" if USE_LOCAL_RESOURCE else "runpod"
    return backend_url(backend)

File: services/meeting-api-rs/src/runtime_backend.rs
//! Replaces the standalone `router` service. That service special-cased `POST /bots` /
//! `DELETE /bots/{platform}/{id}` with a sticky, in-memory local/RunPod split — but nothing
//! ever called those routes (runtime-api only exposes `/containers`, not `/bots`), so real
//! traffic went through its catch-all proxy instead, which routed on a single static
//! `USE_LOCAL_RESOURCE` env var only. There was no actual per-bot load balancing happening.
//!
//! This does what the router's dead code was meant to do, for real: pick a backend once per
//! meeting at spawn time based on current local-bot occupancy vs `LOCAL_BOT_THRESHOLD`, persist
//! the choice in `meetings.data` (survives restarts — the in-memory map didn't), and resolve it
//! back out for every later container operation on that meeting.

use crate::config::Config;
use crate::models::Meeting;
use sqlx::PgPool;

const ACTIVE_STATUSES: &[&str] = &["requested", "joining", "awaiting_admission", "active"];

/// Choose which runtime-api backend a *new* bot should spawn on.
pub async fn choose_backend_for_spawn(db: &PgPool, config: &Config) -> anyhow::Result<&'static str> {
    if !config.use_local_resource {
        return Ok("runpod");
    }
    let local_count: i64 = sqlx::query_scalar(
        "SELECT COUNT(*) FROM meetings \
         WHERE platform != 'browser_session' \
           AND status = ANY($1) \
           AND data->>'runtime_backend' = 'local'",
    )
    .bind(ACTIVE_STATUSES)
    .fetch_one(db)
    .await?;

    Ok(if local_count < config.local_bot_threshold { "local" } else { "runpod" })
}

/// Resolve the already-decided backend for an existing meeting. Falls back to the static
/// `USE_LOCAL_RESOURCE` flag only for meetings created before this field existed.
pub fn backend_url_for<'a>(config: &'a Config, meeting: &Meeting) -> &'a str {
    let backend = meeting.runtime_backend().unwrap_or(if config.use_local_resource { "local" } else { "runpod" });
    backend_url_for_name(config, backend)
}

pub fn backend_url_for_name<'a>(config: &'a Config, backend: &str) -> &'a str {
    if backend == "local" {
        &config.local_backend_url
    } else {
        &config.runpod_backend_url
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn test_config(use_local: bool, threshold: i64) -> Config {
        Config {
            db_host: String::new(),
            db_port: 5432,
            db_name: String::new(),
            db_user: String::new(),
            db_password: String::new(),
            db_pool_size: 1,
            redis_url: String::new(),
            meeting_api_url: String::new(),
            bot_meeting_api_url: String::new(),
            bot_redis_url: String::new(),
            bot_image_name: String::new(),
            local_backend_url: "http://local".to_string(),
            runpod_backend_url: "http://runpod".to_string(),
            use_local_resource: use_local,
            local_bot_threshold: threshold,
            transcription_collector_url: String::new(),
            hivemind_url: String::new(),
            api_keys: vec![],
            internal_api_secret: String::new(),
            dev_mode: false,
            bot_stop_delay_seconds: 90,
            cors_origins: vec![],
        }
    }

    #[test]
    fn backend_url_for_name_picks_local_or_runpod() {
        let config = test_config(true, 3);
        assert_eq!(backend_url_for_name(&config, "local"), "http://local");
        assert_eq!(backend_url_for_name(&config, "runpod"), "http://runpod");
        // anything unrecognized falls to runpod, matching Python's `backend_url()`
        // ("local" if backend == "local" else RUNPOD) exactly
        assert_eq!(backend_url_for_name(&config, "bogus"), "http://runpod");
    }

    #[test]
    fn backend_url_for_falls_back_to_static_flag_when_meeting_has_no_backend_recorded() {
        let config_local_default = test_config(true, 3);
        let config_runpod_default = test_config(false, 3);
        let meeting = Meeting {
            id: 1,
            user_id: 1,
            platform: "google_meet".to_string(),
            platform_specific_id: Some("abc".to_string()),
            status: "active".to_string(),
            bot_container_id: None,
            start_time: None,
            end_time: None,
            data: serde_json::json!({}),
            created_at: None,
            updated_at: None,
        };
        assert_eq!(backend_url_for(&config_local_default, &meeting), "http://local");
        assert_eq!(backend_url_for(&config_runpod_default, &meeting), "http://runpod");
    }
}
